### ISSUE TITLE
zlib header detection is too restrictive.

### ISSUE THREAD

**Issue by madler** [mateidavid/zstr#27]
[This line](https://github.com/mateidavid/zstr/blob/master/src/zstr.hpp#L194) looks only for three specific possible zlib headers. There are many more valid zlib headers that would be missed. 61 to be precise. Look at [RFC 1950](https://tools.ietf.org/html/rfc1950) for the definition of a zlib header.

**Comment by ferdymercury** [comment]
Thanks for the report. Bug confirmed.
See also https://stackoverflow.com/questions/9050260/what-does-a-zlib-header-look-like
If you have a patch, please feel free to submit a PR.

**Comment by madler** [comment]
It looks like you already know to use `inflateInit2()` with `15+32` in order to auto-detect the format. What you're looking for is already built into zlib's inflate.

**Comment by ferdymercury** [comment]
(I am not the author of the software, just an end user / collaborator with admin rights as the original author dropped the project support)

**Comment by dartheditous** [comment]
Not sure if anyone is still around to answer this, but I have a file that starts with 78 5E. I added 78 5E to zstr.hpp's checks to make it attempt to decompress, and it works, but only for the first 2500 bytes or so. Then tellg() starts returning -1 and it no longer reads anything new from the file.

Was 78 53 omitted because it isn't (fully) supported? Or could something else be going wrong?

**Comment by madler** [comment]
No, the `78 5e` zlib header (not `78 53`) does not require any different support. It is likely that your input is corrupted or just not what you're expecting.

**Comment by dartheditous** [comment]
Ah... It Was Windows All Along

zlib's zpipe.c was fine with the file, as was PHP's zlib_decode. But zstr's example zpipe.cpp had to have the following added:

```
  _setmode(_fileno(stdin), _O_BINARY);
  _setmode(_fileno(stdout), _O_BINARY);
```

otherwise it would stop after just a kilobyte or two.

For an ifstream, I have to specify binary mode explicitly:

`  zstr::ifstream is("01 - The Eleventh Hour.mp4.ffindex", std::ios_base::in | std::ios_base::binary);
`

I am admittedly clueless about such things, but shouldn't `ifstream` set the mode to `std::ios_base::binary` by default, as `ofstream` does?

**Comment by ferdymercury** [comment]
> shouldn't `ifstream` set the mode to `std::ios_base::in | std::ios_base::binary` by default,

Related discussion here: https://github.com/mateidavid/zstr/issues/15
